src/Intl/Transliterator/Transliterator.php

```diff
@@ -0,0 +1,676 @@
+            throw new \DomainException(sprintf('The PHP intl extension is required for using "Transliterator->direction".'));
+        }
+
+        return $transliterator;
+    }
+
+    public static function createInverse() {
+        throw new \DomainException(sprintf('The PHP intl extension is required for using "Transliterator::createInverse".'));
+    }
+
+    public static function listIDs() {
+        return array_values(self::$LOCALE_TO_TRANSLITERATOR_ID);
+    }
+
+    public function transliterate($s, $start = null, $end = null) {
+        if ('' === $s) {
+            return '';
+        }
+
+        if (null !== $start) {
+            $s = mb_substr($s, $start, $end);
+        }
+
+        $s = self::clean($s);
+
+        foreach (explode(';', $this->id) as $rule) {
+            $rule = str_replace(array('/BGN', 'ANY-'), '', strtoupper($rule));
+
+            // DEBUG
+            //var_dump($rule);
+
+            if ('NFC' === $rule) {
+                normalizer_is_normalized($s, \Normalizer::FORM_C) ?: $s = normalizer_normalize($s, \Normalizer::NFC);
+            } elseif ('NFD' === $rule) {
+                normalizer_is_normalized($s, \Normalizer::FORM_D) ?: $s = normalizer_normalize($s, \Normalizer::NFD);
+            } elseif ('NFKD' === $rule) {
+                normalizer_is_normalized($s, \Normalizer::FORM_KD) ?: $s = normalizer_normalize($s, \Normalizer::NFKD);
+            } elseif ('NFKC' === $rule) {
+                normalizer_is_normalized($s, \Normalizer::FORM_KC) ?: $s = normalizer_normalize($s, \Normalizer::NFKC);
+            } elseif ('[:NONSPACING MARK:] REMOVE' === $rule) {
+                $s = preg_replace('/\p{Mn}++/u', '', $s);
+            } elseif (false !== strpos($rule, 'REMOVE')) {
+                $s = preg_replace(self::$REGEX_ASCII, '', $s);
+            } elseif (false !== strpos($rule, 'LATIN-ASCII')) {
+                $s = self::to_ascii($s);
+            } elseif (false !== strpos($rule, 'UPPER')) {
+                $s = mb_strtoupper($s);
+            } elseif (false !== strpos($rule, 'LOWER')) {
+                $s = mb_strtolower($s);
+            } elseif (false !== strpos($rule, 'LATIN')) {
+                $s = self::to_translit($s);
+            } elseif ($lang = array_search($rule, self::$LOCALE_TO_TRANSLITERATOR_ID)) {
+                $s = self::to_ascii($s, $lang);
+            } elseif (
+                false !== strpos($rule, '-ASCII')
+                &&
+                $lang = str_replace('-ASCII', '', $rule)
+            ) {
+                $s = self::to_ascii($s, $lang);
+            }
+        }
+
+        return $s.(null !== $start ? mb_substr($s, $end) : '');
+    }
+    
+    private static function to_translit($s) {
+        if (null === self::$TRANSLIT) {
+            $array = self::getData('translit');
+
+            self::$TRANSLIT = array(
+                'orig' => array_keys($array),
+                'replace' => array_values($array),
```

Comment:
It would be great to have a source data format that doesn't require any processing, as that would allow using shared memory directly.

Reply:
done